src/mongo_logger.ts

```diff
@@ -0,0 +1,214 @@
+  MONGODB_LOG_COMMAND?: string;
+  /** Severity level for topology component */
+  MONGODB_LOG_TOPOLOGY?: string;
+  /** Severity level for server selection component */
+  MONGODB_LOG_SERVER_SELECTION?: string;
+  /** Severity level for CMAP */
+  MONGODB_LOG_CONNECTION?: string;
+  /** Default severity level to be if any of the above are unset */
+  MONGODB_LOG_ALL?: string;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  MONGODB_LOG_MAX_DOCUMENT_LENGTH?: string;
+  /** Destination for log messages. Must be 'stderr', 'stdout'. Defaults to 'stderr'. */
+  MONGODB_LOG_PATH?: string;
+}
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  /** Destination for log messages */
+  mongodbLogPath?: 'stdout' | 'stderr' | Writable;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  componentSeverities: {
+    /** Severity level for command component */
+    command: SeverityLevel;
+    /** Severity level for topology component */
+    topology: SeverityLevel;
+    /** Severity level for server selection component */
+    serverSelection: SeverityLevel;
+    /** Severity level for connection component */
+    connection: SeverityLevel;
+    /** Default severity level to be used if any of the above are unset */
+    default: SeverityLevel;
+  };
+
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. */
+  logDestination: Writable;
+}
+
+/**
+ * Parses a string as one of SeverityLevel
+ *
+ * @param s - the value to be parsed
+ * @returns one of SeverityLevel if value can be parsed as such, otherwise null
+ */
+function parseSeverityFromString(s?: string): SeverityLevel | null {
+  const validSeverities: string[] = Object.values(SeverityLevel);
+  const lowerSeverity = s?.toLowerCase();
+
+  if (lowerSeverity != null && validSeverities.includes(lowerSeverity)) {
+    return lowerSeverity as SeverityLevel;
+  }
+
+  return null;
+}
+
+/**
+ * Parses a string to be a number greater than or equal to 0 for maxDocumentLength.
+ *
+ * @param s - the value to be parsed
+ * @returns the int value parsed or 1000 if the value could not be parsed
+ */
+function parseMaxDocumentLength(s?: string): number {
+  try {
+    const maxDocumentLength = getUint('MONGODB_LOG_MAX_DOCUMENT_LENGTH', s);
+    return maxDocumentLength;
+  } catch {
+    return 1000;
+  }
+}
+
+/**
+ * resolves the MONGODB_LOG_PATH and mongodbLogPath options from the environment and the
+ * mongo client options respectively.
+ *
+ * @returns the Writable stream to write logs to
+ */
+function resolveLogPath(
+  { MONGODB_LOG_PATH }: MongoLoggerEnvOptions,
+  {
+    mongodbLogPath
+  }: {
+    mongodbLogPath?: unknown;
+  }
+): Writable {
+  const isValidLogDestinationString = (destination: string) =>
+    ['stdout', 'stderr'].includes(destination.toLowerCase());
```

Comment:
This is fine, but it did catch my attention as slightly unusual structuring, specifically creating a new function for a condition check.

I'd expect `['stdout', 'stderr']` to but pulled up to a module variable and then each if stmt below to call `VALID_DESTS.includes(destination.toLowerCase())`

Reply:
Interesting, this is the approach I always use.  

1. no reason for top-level variables if they're used in one place, so place them in the function to not pollute the module's scope
2. use descriptive names

but i don't feel strongly, I can change if you want

Reply:
No worries, just caught my eye as different, all good 👍